clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/AddExecutor.scala

```diff
@@ -42,8 +46,106 @@ class AddExecutor[TI <: TransferIndex](addCommand: AddCommand[TI])
       Future
         .failed(_) logErrorMsg s"Metadata at $location cannot be added to Clio", {
         decoded =>
-          webClient.upsert(addCommand.index)(addCommand.key, decoded)
+          {
+            for {
+              existingMetadata <- queryForKey(webClient).logErrorMsg(
+                s"Could not query the $name. No files have been deleted."
+              )
+              upsertResponse <- addFiles(
+                webClient,
+                decoded,
+                existingMetadata,
+                addCommand.forceUpdate
+              ).fold(
+                left => Future.failed(left) logErrorMsg left.getMessage,
+                right => right
+              )
+            } yield {
+              upsertResponse
+            }
+          }
       }
     )
   }
+
+  private def addFiles(
+    client: ClioWebClient,
+    newMetadata: addCommand.index.MetadataType,
+    existingMetadata: Option[addCommand.index.MetadataType],
+    forceUpdate: Boolean
+  )(implicit ec: ExecutionContext): Either[Throwable, Future[UpsertId]] = {
+    val map = existingMetadata.map(diffMetadata(_, newMetadata)).getOrElse(Map.empty)
+    for {
+      upsertResponse <- Either.cond(
+        map.isEmpty || forceUpdate,
+        client
+          .upsert(addCommand.index)(addCommand.key, newMetadata)
+          .logErrorMsg(
+            s"""An error occurred while adding the $name record in Clio.""".stripMargin
+          ),
+        new RuntimeException(
+          s"Adding this document that will overwrite the following existing metadata: " +
+            map.keys.mkString(", ") +
```

Comment:
Hrm, would it be valuable to include the existing values on top of the keys? Also this seems like it could build up a huge line, would it make sense to join with `"\n"` instead of `", "`?

Reply:
Yeah I debated this. Unfortunately the diff returns the fieldNames and values of the newMetadata. I can run the diff in both directions though and tie them together with the field name. I'll give it a shot.